### PR TITLE
Fix flakiness with kubectl wait tests

### LINKED ISSUE #111118: kubectl wait tests flaking after merging #110923
### Which jobs are flaking?

Related to https://github.com/kubernetes/kubernetes/pull/110923

See more info at https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=kubectl%2Fpkg%2Fcmd%2Fwait

### Which tests are flaking?

List of flaking tests in the job:

- TestWaitForJSONPathCondition
- TestWaitForJSONPathCondition/handles_watch_condition_change
- TestWaitForCondition
- TestWaitForJSONPathCondition/handles_watch_created
- TestWaitForCondition/times_out_due_to_stale_.status.conditions[0].observedGeneration
- TestWaitForCondition/handles_watch_created

### Since when has it been flaking?

July 7, 2022

### Testgrid link

_No response_

### Reason for failure (if possible)

Recent changes in `kubectl wait` subcommand, unsure of exact root cause as of this writing but will update with that information as I find it.

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig cli

### PR DESCRIPTION
**Pull request by mpuckett159**
#### What type of PR is this?
/kind flake

#### What this PR does / why we need it:
https://github.com/kubernetes/kubernetes/pull/110923 introduced some test flakiness. This PR seeks to resolve this. The issue was a result of PrependWatchReactor not always creating the necessary object to be watched before the watch test started watching, resulting in an "object not found" error. This has been resolved by switching to PrependReactor, which appears to force the object creation before the watch is started, ensuring it is present when the watch is needed.

`TestWaitForJSONPathCondition/handles_watch_created` - I believe this was a result of PrependWatchReactor not necessarily taking affect before the first watch task happens, causing the watch to fail due to the missing resource. I switched to using PrependReactor which reduced failures from 1/200-300 to not being observed in over 1300 tests.

`TestWaitForJSONPathCondition/handles_watch_condition_change` - This was similar to `handles watch created` in that it needed the use of PrependReactor instead of PrependWatchReactor.

`TestWaitForCondition/handles_watch_created` - Same issue as the `TestWaitForJSONPathCondition` version.

`TestWaitForCondition/times_out_due_to_stale_.status.conditions[0].observedGeneration` - this one appeared to be the result of the possibility of the stale observed generation happening before there were two wait actions completed. To fix this I allowed for the wait list length to be between 1 and 2 actions in length. This decreased flakiness to the point that I was no longer able to observe it anymore in more than 1300 tests.

`TestWaitForCondition/times_out` - not originally listed but similar issue as observed generation stale test. Allowed between 1 and 2 actions in list.

`TestWaitForDeletion/handles_watch_delete_multiple` - not originally listed but similar to observed generation stale test. Allow between 4 and 6 actions in list.

#### Special note for reviewers
The root cause for these flakes is, I believe, the introduction of the informers, which use a precondition function to do an initial "get", which can lead some odd scenarios where testing environments can not be completely finished initializing themselves, but the precondition function is running a get against them and seeing that the expected resource does not exist.

This also affected the observed generation error test because it lead to there potentially only being a single action in the list because the precondition function would burn one of the two expected actions to be taken. That is why it was necessary to allow for a list length of either 1 or 2 actions.

#### Which issue(s) this PR fixes:
Fixes #111118

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
+1 on reverting rather than adding timeouts/retries if we don't understand the root issue

**Comment by mpuckett159** [comment]
Ok, that's fine, I'm going to continue to investigate then, I believe I know the root cause but I'm just not familiar with how to "solve" the problem. Based on my testing it seems that the channel that is used to monitor the wait doesn't get set up properly, I originally thought it was a timing issue but adding a simple sleep didn't help the issue unfortunately. I think this is specifically a testing issue because I tried recreating it by just spamming `kubectl wait` calls for about 30 minutes and was never able to recreate the failure. 

I'll try reaching out to someone on the sig-cli with more experience than me and see if we can't figure out a better solution.

/hold

**Comment by mpuckett159** [comment]
Alright this should reduce the flakiness dramatically. It seems like the `PrependWatchReactor` can't be relied on to get the resource we're testing on added into the fake client before we do our initial get function to populate the cache for the waiter. Additionally, it looks like the movement to informers lead to some variability in what to expect for the list of actions taken by the client.

I'm not sure if there should be any changes made to the `k8s.io/client-go/dynamic/fake` library because of this. I can try poking around I guess.

File: staging/src/k8s.io/kubectl/pkg/cmd/wait/wait_test.go
Comment (aojea) [inline staging/src/k8s.io/kubectl/pkg/cmd/wait/wait_test.go:578]: how can len(actions) be less than 5 and greater than 6 at the same time :smile: ?

**Comment by soltysh** [comment]
+1 to revert to unblock the queue

@mpuckett159 looking through the tests again, how about making them less strict about the contents of actions array and rather look for existence of particular actions rather than existence at a specific index. This will ensure that the actions happened, but not necessarily bind to specific order, which is not given with informers.

**Comment by mpuckett159** [comment]
@soltysh to be honest the more I've been looking at this the less I think the tests of the actions themselves matter. Since we're moving to a less strict system with the informers and especially with the precondition function, it doesn't really matter that we have a list action, then a get action, then a watch action, etc, so if it's alright with you I think I might just axe that pattern all together.

I'll get a PR together to revert the original change again asap. It looks like there's another potential issue being caused by it as well.

**Comment by mpuckett159** [comment]
Please excuse my floundering around with git I've never done a cherry pick before and really botched it lol

File: staging/src/k8s.io/kubectl/pkg/cmd/wait/wait_test.go
Comment (mpuckett159) [reply staging/src/k8s.io/kubectl/pkg/cmd/wait/wait_test.go:578]: Whoops, thanks for catching that. I think I'm just gonna strip out the validate action testing honestly. With the move to informers the number of actions taken becomes variable, and the number of potential valid cases to test for spirals out of control very fast.

**Comment by mpuckett159** [comment]
Note to any reviewers, I am still waiting for more information on another issue that the original PR may have introduced. The current state of these changes ran close to 15k tests with no failures, but I still need to track down this other issue because I've been unable to observe it directly.

**Comment by mpuckett159** [comment]
> Note to any reviewers, I am still waiting for more information on another issue that the original PR may have introduced. The current state of these changes ran close to 15k tests with no failures, but I still need to track down this other issue because I've been unable to observe it directly.

I believe that the failures observed and reported are still persisting in the scalability test, after the original changes were reverted, so I think we can safely assume my changes were not the cause, but waiting for confirmation still anyway.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/label tide/merge-method-squash

**Comment by soltysh** [comment]
/priority backlog
/hold cancel
from my current tests and based on the investigations I think this should be safe to merge

**Pull request merged**
